planning/behavior_velocity_planner/include/scene_module/dynamic_obstacle_stop/utils.hpp

```diff
@@ -0,0 +1,227 @@
+
+#include <vehicle_info_util/vehicle_info_util.hpp>
+
+#include <string>
+#include <vector>
+
+namespace behavior_velocity_planner
+{
+namespace dynamic_obstacle_stop_utils
+{
+namespace bg = boost::geometry;
+using autoware_auto_perception_msgs::msg::ObjectClassification;
+using autoware_auto_perception_msgs::msg::PredictedObjects;
+using autoware_auto_planning_msgs::msg::PathWithLaneId;
+using tier4_autoware_utils::Box2d;
+using tier4_autoware_utils::LineString2d;
+using tier4_autoware_utils::Point2d;
+using tier4_autoware_utils::Polygon2d;
+using tier4_debug_msgs::msg::Float32Stamped;
+using vehicle_info_util::VehicleInfo;
+using PathPointsWithLaneId = std::vector<autoware_auto_planning_msgs::msg::PathPointWithLaneId>;
+struct CommonParam
+{
+  double normal_min_jerk;  // min jerk limit for mild stop [m/sss]
+  double normal_min_acc;   // min deceleration limit for mild stop [m/ss]
+  double limit_min_jerk;   // min jerk limit [m/sss]
+  double limit_min_acc;    // min deceleration limit [m/ss]
+};
+struct DynamicObstacleStopParam
+{
+  std::string detection_method;
+  bool use_partition_lanelet;
+  bool specify_decel_jerk;
+  double stop_margin;
+  double passing_margin;
+  double deceleration_jerk;
+  double obstacle_velocity_kph;
+  float detection_distance;
+  float detection_span;
+  float min_vel_ego_kmph;
+};
+
+struct VehicleParam
+{
+  float base_to_front;
+  float base_to_rear;
+  float width;
+};
+
+struct DetectionAreaSize
+{
+  float dist_ahead;
+  float dist_behind;
+  float dist_right;
+  float dist_left;
+};
+
+struct ApproachingParam
+{
+  bool enable;
+  float margin;
+  float limit_vel_kmph;
+  float stop_thresh;
+  float stop_time_thresh;
+  float dist_thresh;
+};
+
+struct SlowDownLimit
+{
+  bool enable;
+  float max_jerk;
+  float max_acc;
+};
+
+struct PlannerParam
+{
+  CommonParam common;
+  DynamicObstacleStopParam dynamic_obstacle_stop;
+  VehicleParam vehicle_param;
+  DetectionAreaSize detection_area;
+  ApproachingParam approaching;
+  DynamicObstacleParam dynamic_obstacle;
+  SlowDownLimit slow_down_limit;
+};
+
+enum class State {
+  GO = 0,
+  APPROACH,
+  STOP,
+};
+
+enum class DetectionMethod {
+  Object = 0,
+  ObjectWithoutPath,
+  Points,
+  Unknown,
+};
+
+bool validCheckDecelPlan(
+  const double v_end, const double a_end, const double v_target, const double a_target,
+  const double v_margin, const double a_margin);
+
+/**
+ * @brief calculate distance until velocity is reached target velocity (TYPE1)
+ * @param (v0) current velocity [m/s]
+ * @param (vt) target velocity [m/s]
+ * @param (a0) current acceleration [m/ss]
+ * @param (am) minimum deceleration [m/ss]
+ * @param (ja) maximum jerk [m/sss]
+ * @param (jd) minimum jerk [m/sss]
+ * @param (t_min) duration of constant deceleration [s]
+ * @return moving distance until velocity is reached vt [m]
+ * @detail TODO(Satoshi Ota)
+ */
+boost::optional<double> calcDecelDistPlanType1(
```

Comment:
[Q] what is DistPlanType1,2,3 ? can you name more concretely? and if possible can you make a util for new deceleration method or replace this to planning_utils?

Reply:
Thank you for your review!

They are calculations to decelerate with constant jerk. Depending on the initial velocity and acceleration, three different calculations are required.
I will create another PR and move these functions to util, and add more detailed explanations. 
(I would like to modify obstacle_stop_planner to use this util, since it uses the same function as well.
https://github.com/autowarefoundation/autoware.universe/blob/main/planning/obstacle_stop_planner/src/node.cpp#L89-L229

Reply:
good and I think this kind of [hard coded](https://github.com/autowarefoundation/autoware.universe/blob/f2577c6f2193578894beb3846446dcbee7d59e16/planning/obstacle_stop_planner/src/node.cpp#L123
) params needs explanation because there are some other [margins](https://github.com/autowarefoundation/autoware.universe/blob/f2577c6f2193578894beb3846446dcbee7d59e16/planning/obstacle_stop_planner/src/node.cpp#L230) as well and it's difficult to distinguish which margin is for what.

Reply:
okay, I will add explanations for them when move them to utils.